Set statsig_stable_id cookie based on OneTrust consent

nicklathe opened the pull request:
This PR re-introduces setting the `statsig_stable_id` as a cookie in order to support a shared ID across subdomains. This is so a signed-out user who navigates across studio.code.org and code.org can be tracked as a single user. This also supports the "Last Touch" initiative, allowing UTM's evented from code.org to be later tied to users on studio.code.org via the stable id.

This functionality also introduces logic to only set this cookie based on a user's OneTrust consent, i.e. it will remove the cookie if the user has disabled performance cookies.

Lastly, this functionality honors a cookie set statsig_stable_id on the backend, adding it to the session if it exists for server side events, and allowing the backend to generate it's own if the cookie is not set.

### Testing Locally

* While running locally, navigate between `localhost.code.org` and `localhost-studio.code.org`. The cookie `statsig_stable_id` should remain stable between these domains
* In the footer, click the `Cookie Notice` link. From there, you can click `Manage Cookies`. Disable `Performance Cookies`
* Navigate back to `localhost-studio.code.org`. After the page fully renders, the cookie should no longer be set
* Re-enable the `Performance Cookie`
* Navigate back to code studio. After the page fully renders, the cookie should be set again
* Sign out, and the cookie will be deleted (but quickly repopulated once you're redirected to `localhost.code.org` again
* To see a server side event with the `statsig_stable_id`, create a new user. At the end of this flow, you should see the `Sign Up Finished Backend` event in your terminal. The `statsig_stable_id` should match the one set in the cookie

## Overview
<img width="1475" height="915" alt="Screenshot 2025-10-27 at 2 33 05 PM" src="https://github.com/user-attachments/assets/7433ab1f-5910-4d65-8caf-a14b44f2aa12" />

## Screen Shots

#### Land on Code.org, statsig_stable_id cookie is set
<img width="1217" height="706" alt="Screenshot 2025-10-23 at 5 37 07 PM" src="https://github.com/user-attachments/assets/ca2b19ab-f387-4f34-9b2b-93390e1b6412" />

#### Login to studio.code.org, statsig_stable_id remains the same cross domain
<img width="1216" height="702" alt="Screenshot 2025-10-23 at 5 38 09 PM" src="https://github.com/user-attachments/assets/ce623ae1-9172-40f6-a5c1-22ecb93bd46d" />

#### statsig_stable_id is propagated to the backend
<img width="344" height="86" alt="Screenshot 2025-10-23 at 5 40 30 PM" src="https://github.com/user-attachments/assets/ecdb8738-6270-4ab4-ba8f-28a70b457e55" />

#### OneTrust consent changed to not allow performance cookies
<img width="475" height="418" alt="Screenshot 2025-10-23 at 5 41 07 PM" src="https://github.com/user-attachments/assets/8d5bcbf7-1f0d-4802-af3c-922f63ae918d" />

#### Stable ID cookie removed (but still set in session for server-side logging)
<img width="1217" height="706" alt="Screenshot 2025-10-23 at 5 41 33 PM" src="https://github.com/user-attachments/assets/0925c7d3-c9e0-4d0f-be02-160fe2c5d1af" />

#### OneTrust consent re-enabled for performance cookies
<img width="501" height="415" alt="Screenshot 2025-10-23 at 5 42 44 PM" src="https://github.com/user-attachments/assets/3e7c9389-319b-4fa3-9250-037dafe613cc" />

#### New Stable ID cookie set
<img width="1212" height="704" alt="Screenshot 2025-10-23 at 5 43 18 PM" src="https://github.com/user-attachments/assets/dc61e23f-d4e6-4b58-95f8-86ed0eafa659" />

#### New Stable ID propagated to backend
<img width="338" height="64" alt="Screenshot 2025-10-23 at 5 43 28 PM" src="https://github.com/user-attachments/assets/e43ec367-ffc2-4cea-b03e-7fa57bb4efdb" />

#### After signing out, cookie is cleared out. Redirected to code.org and new cookie is set
<img width="1218" height="711" alt="Screenshot 2025-10-23 at 5 44 12 PM" src="https://github.com/user-attachments/assets/a05a43ac-173f-43af-a4d2-7816a5aeb978" />


## Links
<!--  Links to relevant external resources.
      - spec docs, Jira tickets, related PRs, Honeybadger errors, etc.
-->

- [Last Touch proposal](https://docs.google.com/document/d/1n_qx0iJjQOn0mIjcjkUt2x6wIvO15UlwxJKGDO8FiO0/edit?tab=t.fbguugq0n3jl#heading=h.vysvfanbwbam)

## Testing story

## Follow-up work
In order to not break OneTrust, we will need to initiate a new scan and categorize this cookie as a performance cookie

nicklathe committed:
Revert "Stop setting statsig_stable_id as cookie (#65673)"

  - This PR re-introduced setting the statsig_stable_id as a cookie
    in order to support a shared ID across subdomains.

Signed-off-by: Nick Lathe <nick.lathe@code.org>

carl-codeorg commented on apps/src/metrics/statsigHelpers.js:
```diff
@@ -10,9 +12,16 @@ export function getUserType() {
-  const scriptTag = document.querySelector('script[data-statsig-stable-id]');
-  return scriptTag?.dataset?.statsigStableId || null;
+export function findOrCreateStableId() {
+  const STABLE_ID_KEY = 'statsig_stable_id';
```
Should this be put in `AnalyticsConstants.js`?

carl-codeorg reviewed (approved):
LGTM

nicklathe committed:
Set cookie based on OneTrust consent

  - Add logic to wait for OneTrust javascript to load, and set/remove
    statsig_stable_id cookie conditionally based on consent
  - Refactor backend statsig_stable_id logic to set based on cookie
    value if present, and fallback to it's own if not present

nicklathe committed:
Fix StudioReporter so it doesn't error in unit tests

stephenliang commented on apps/src/metrics/statsigHelpers.js:
```diff
@@ -27,3 +74,66 @@ export function formatUserId(userId) {
     return `${environment}-${userIdString}`;
   }
 }
+
```
This shouldn't be needed, there's a OneTrust callback that informs when OneTrust is ready. To convert to a promise, we can assign a global `window.oneTrustPromise` (preferably in the header after the OneTrust script tag).

```
      window.oneTrustPromise = new Promise(function (resolve) {
          window.OptanonWrapper = function OptanonWrapper() {
            resolve(window.OneTrust);
          };
        });
```

stephenliang commented on apps/src/metrics/statsigHelpers.js:
```diff
@@ -10,9 +28,38 @@ export function getUserType() {
-  const scriptTag = document.querySelector('script[data-statsig-stable-id]');
-  return scriptTag?.dataset?.statsigStableId || null;
+export function findOrCreateStableId() {
+  let stableId = cookies.get(STABLE_ID_KEY);
+
+  if (!stableId) {
+    stableId = createUuid();
+  }
+
+  if (consentAllowsStatsigCookie()) {
+    cookies.set(STABLE_ID_KEY, stableId, COOKIE_OPTIONS);
+  } else {
+    // Ensure any existing cookie is removed to satisfy OneTrust
+    // (must pass same attributes used when setting the cookie)
+    cookies.remove(STABLE_ID_KEY, {path: '/', domain: '.code.org'});
+  }
+
+  return stableId;
+}
+
+function consentAllowsStatsigCookie() {
+  const groups = getOnetrustGroups();
+  return ONETRUST_ALLOWED_CATEGORIES.some(id => groups.has(id));
+}
+
+function getOnetrustGroups() {
+  try {
+    const rawString =
+      typeof window !== 'undefined' ? window.OnetrustActiveGroups || '' : '';
+    // console.log(`Onetrust groups: ${rawString}`);
+    return new Set(rawString.split(',').filter(Boolean));
+  } catch (error) {
+    return new Set();
+  }
```
I'd like to see a different strategy to hoepfully simplify the auto block and scanning process going forward, and reduce custom code to implement this:

1. Create a new webpack entrypoint for Statsig analytics including this cookie setting
2. Include this script after OneTrust
3. Autoblock will then block this webpack entrypoint (good use case)
4. Ensure that code that references the entrypoint works even if entrypoint is blocked.

We can do this as a follow up since I feel like Statsig is a good candidate for modularization and we can probably roll that in to that work if you didn't want to wait.

nicklathe replied:
I agree, I was thinking along a similar line, that this should be some kind of package that could be used in both dashboard and the marketing site. I'd like to still get this merged, as there are other stakeholders who are relying on this functionality, but agree it's a good candidate for refactor down the road.

daynew reviewed (approved):
Read through the code and this addresses all the concerns I can think of.

nicklathe committed:
PR feedback and bug fix

  * Address PR feedback by using OneTrust callback instead of custom
    function
  * Fix bug where options.environment was getting overwritten

nicklathe committed:
Fix test
